### ISSUE TITLE
Query: Should getCurrentPosition do a HTTPRequest if it's options have persist:true

### ISSUE THREAD

**Issue by littleowlnest** [transistorsoft/cordova-background-geolocation-lt#836]
## Your Environment
* Plugin version: 2.14.2
* Platform: Android
* OS version: Android One
* Device manufacturer / model: Xiaomi Mi A2
* Cordova version (`cordova -v`): 8.1.2 
* Cordova platform version (`cordova platform ls`): Android 7.1.3
* Plugin config

                    desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
                    distanceFilter: 10,
                    allowIdenticalLocations: true,
                    heartbeatInterval: 60,
                    stationaryRadius: 25,
                    startOnBoot: true,
                    stopOnStationary: false,
                    stopTimeout: 1,         // Change state to stationary after 1 min with device "still"
                    stopOnTerminate: false, // Don't stop tracking when app is terminated.
                    foregroundService: true,
                    persist: true,
                    desiredOdometerAccuracy: 100,
                    logLevel: BackgroundGeolocation.LOG_LEVEL_VERBOSE,
                    preventSuspend: true,
                    autoSync: true,
                    autoSyncThreshold: 3,
                    batchSync: true,
                    maxBatchSize: 3,
                    url: window.serviceBase + 'api/tracking/sync',
                    debug: true,  // <-- Debug sounds & notifications.
                    

## Expected Behavior
Heartbeat is triggering correctly every minute, execute a getCurrentPosition that has the option set to persist.  I expect the plugin to execute a getCurrentPosition and store in the local SQLLite db.  Since autosync is setup, I would then expect it to sync to the server URL.

## Actual Behavior
Heartbeat is triggered and getCurrentPosition is called but no HTTPRequest is made.

## Steps to Reproduce

1. Here is the code being executed:

   bgGeo.on('heartbeat', function (params) {
                    var options = {
                        timeout: 60,
                        persist: true,        
                        maximumAge: 5000,     
                        desiredAccuracy: BackgroundGeolocation.DESIRED_ACCURACY_HIGH,
                        samples: 3         
                    };

                    var position = BackgroundGeoLocation.getCurrentPosition(options);
                    
                });

## Context
Yes - I have already read the Philosophy of Operation.  The use case I have is that am writing an app to manage reps out in the field that are allocated to be zoned or free roaming.  Free roaming means they're not supposed to stop moving for too long.  This plugin works perfectly for this scenario.  Zoned means that they mill around a particular area, sometimes standing or sitting in a mall.

Serverside I receive requests that basically do a "check-in" to confirm that they're still in the location they said they were.  The "check-in" prevents them from turning off their phone for 30 minutes or so and disappearing off to do whatever they want - returning and turning their phone back on as if they were sitting there the whole time.

At the moment I'm not receiving the check-ins when the person is stationary because no HTTPRequest is being made.

Theoretically, a quick fix is to set stopTimeout to some huge value.  Seems like a waste though.

## Debug logs
Not sure logs help at the moment if the plugin doesn't handle this use case.

**Comment by christocracy** [comment]
Answer is *yes*.

Look at your logs.  You’ll see plugin attempting http request in the logs.

**Comment by christocracy** [comment]
And do you understand what this does?

autoSyncThreshold: 3,

**Comment by littleowlnest** [comment]
Queues up 3 records before syncing.  So theoretically if it has been sitting for several minutes, it should end up sending at least once or twice a batch of 3.... I'm hoping.

**Comment by christocracy** [comment]
Show me `$ adb logcat -s TSLocationManager` while your heartbeat event fires.

**Comment by littleowlnest** [comment]
BTW... before we go any further.  Your support of the plugin is top notch.
I think you were right about the batching.  I wasn't waiting long enough, I turned off batching and I got a single response.  (There is a JSON serialization error... but that's on me).

So... here's another question, can I configure it to use batching and then onStationary use setConfig to turn it off?  And then onMoving turn it back on?

**Comment by christocracy** [comment]
> can I configure it to use batching and then onStationary use setConfig to turn it off?

Of course.  You're free to `#setConfig` anything at runtime.

You can also simply call `#sync` at any time.  `#sync` overrides `autoSyncThreshold`

```javascript
BackgroundGeolocation.getCurrentPosition(options).then((location) => {
  BackgroundGeolocation.sync();
});
```

**Comment by littleowlnest** [comment]
Sync sounds like a much more elegant solution.

**Comment by littleowlnest** [comment]
Sorry... wasn't sure if I should log a new issue.  But basically, I've noticed if you do batches, it will send it as an array which is enclosed with square braces [].  A single request will not have square braces.  This generally screws with ModelBinding server side, since it doesn't know if it's receiving an array or not.

Theoretically, I can just process the raw JSON and determine if it is an array or not by checking for the braces and then deserialize appropriately, but it's kind of a hack.

**Comment by christocracy** [comment]
I see no issue here iOS or Android.  With `batchSync: true`, the plugin's HTTP service will always post an `[ ]`, even if it contains only a single location.

**iOS**
```
POST /locations
{
  "host": "192.168.11.12:9000",
  "content-type": "application/json",
  "connection": "keep-alive",
  "accept": "*/*",
  "user-agent": "BG%20Geo/2.14.0 CFNetwork/975.0.3 Darwin/18.2.0",
  "content-length": "592",
  "accept-language": "en-us",
  "accept-encoding": "gzip, deflate"
}
Authorization: undefined
{
  "location": [   // <-- Array
    {
      "coords": {
        "speed": -1,
        "longitude": -73.6166638042786,
        "floor": null,
        "latitude": 45.51924196013187,
        "accuracy": 32,
        "altitude_accuracy": 3,
        "altitude": 74.7,
        "heading": -1
      },
      "extras": {
        "foo": "bar"
      },
      "is_moving": false,
      "odometer": 24902.6,
      "uuid": "30E64CEB-E8E0-4084-A6DF-C5EDA0A6B074",
      "activity": {
        "type": "still",
        "confidence": 100
      },
      "battery": {
        "level": 0.9800000190734863,
        "is_charging": true
      },
      "timestamp": "2018-11-26T15:45:59.688Z"
    }
  ]  
}
```

**Android**
```
POST /locations
{
  "content-type": "application/json; charset=utf-8",
  "content-length": "481",
  "host": "192.168.11.12:9000",
  "connection": "Keep-Alive",
  "accept-encoding": "gzip",
  "user-agent": "okhttp/3.10.0"
}
Authorization: undefined
{
  "location": [ // <-- Array
    {
      "is_moving": false,
      "uuid": "e6cf0bfc-7590-4071-9965-a089c434f5d3",
      "timestamp": "2018-11-26T15:47:58.906Z",
      "odometer": 20549.2,
      "coords": {
        "latitude": 45.5192389,
        "longitude": -73.6170423,
        "accuracy": 15,
        "speed": -1,
        "heading": -1,
        "altitude": 46.2
      },
      "activity": {
        "type": "still",
        "confidence": 100
      },
      "battery": {
        "is_charging": true,
        "level": 1
      },
      "extras": {
        "foo": "bar"
      }
    }
  ]
}
```

**Comment by littleowlnest** [comment]
Hi - thanks for the thorough response.  I realised after reading your last comment that my testing was inconsistent.  I was looking at the HTTP Request from when I was doing the batch processing and then I turned it off for the getCurrentPosition call in the heartbeat event, which didn't have batch processing, which obviously makes it inconsistent. (I do recall you closing an issue some time ago related to the braces).  Consider this a non-issue.
